### PR TITLE
Type hints

### PR DESCRIPTION
**Pull request by p770193**
Type hints for better user experience with IDE.

confer https://docs.python.org/3/library/typing.html

- [x] get method output type hints:  Data, Axis, Variable, Channel, Group
- [ ] ~~get method input type hints~~
- [ ] ~~support type hints for natural naming of children~~

### CODE REVIEW

**Comment by p770193** [comment]
@ksunden can you think of a good way to type hint e.g. for `data.w1`?

**Comment by ksunden** [comment]
@p770193, No, there really is no good way, as those are purely dynamic names (and in fact can, and do, change what type they are during execution)

Once that line is executed, it will know the type (not just a hint) and things like tab completion will work in the console. further, it will not autocomplete prior to the first time you try to accesses things by natural name, but will after that first time (at least in the console, not sure about the IDE text editors)

**Comment by ksunden** [comment]
Further, those are accessed using the `__getattr__` method, which can return many different things, Axes, variables, channels for data objects (possibly even collections/bare hdf5 datasets/groups, though that is not recommended if it can be avoided)

File: WrightTools/data/_channel.py [hunk 1]
```diff
@@ -8,11 +8,13 @@
...
 import h5py
 
+from typing import TypeVar
+
 from .. import kit as wt_kit
 from .._dataset import Dataset
 
 __all__ = ["Channel"]
-
+num_type = TypeVar("num_type", int, float, complex)
```
Comment (ksunden) [inline WrightTools/data/_channel.py]: Perhaps for future, but we could consider making channel, variable, etc, Generic types. Personally, wish to wait to see how numpy/h5py handle such things

File: WrightTools/data/_data.py [hunk 2]
```diff
@@ -621,7 +621,7 @@ def create_variable(self, name, values=None, shape=None, units=None, **kwargs):
         self.attrs["variable_names"] = np.append(self.attrs["variable_names"], name.encode())
         return variable
 
-    def downscale(self, tup, name=None, parent=None):
+    def downscale(self, tup, name=None, parent=None) -> "Data":
```
Comment (ksunden) [inline WrightTools/data/_data.py]: What's up with the string?

File: WrightTools/data/_data.py [hunk 3]
```diff
@@ -673,8 +678,8 @@ def downscale(self, tup, name=None, parent=None):
     def flush(self):
         super().flush()
 
-    def get_nadir(self, channel=0):
-        """Get the coordinates in units of the minimum in a channel.
+    def get_nadir(self, channel=0) -> tuple:
```
Comment (ksunden) [inline WrightTools/data/_data.py]: typing.Tuple?

File: WrightTools/data/_data.py [hunk 4]
```diff
@@ -699,8 +704,8 @@ def get_nadir(self, channel=0):
         # finish
         return tuple(a[idx] for a in self._axes)
 
-    def get_zenith(self, channel=0):
-        """Get the coordinates in units of the maximum in a channel.
+    def get_zenith(self, channel=0) -> tuple:
```
Comment (ksunden) [inline WrightTools/data/_data.py]: typing.Tuple?

File: WrightTools/data/_data.py [hunk 5]
```diff
@@ -827,7 +832,7 @@ def level(self, channel, axis, npts, *, verbose=True):
 
     def map_variable(
         self, variable, points, input_units="same", *, name=None, parent=None, verbose=True
-    ):
+    ) -> "Data":
```
Comment (ksunden) [inline WrightTools/data/_data.py]: string?

File: WrightTools/data/_data.py [hunk 6]
```diff
@@ -1244,7 +1249,7 @@ def f(_, s, channels):
 
         self.channels[0].chunkwise(f, self.channels)
 
-    def smooth(self, factors, channel=None, verbose=True):
+    def smooth(self, factors, channel=None, verbose=True) -> "Data":
```
Comment (ksunden) [inline WrightTools/data/_data.py]: String?

File: WrightTools/kit/_array.py [hunk 7]
```diff
@@ -6,6 +6,8 @@
 
 import numpy as np
 
+from typing import Tuple, List, Union, Any
```
Comment (ksunden) [inline WrightTools/kit/_array.py]: List, Union unused

File: WrightTools/kit/_array.py [hunk 8]
```diff
@@ -310,7 +312,7 @@ def unique(arr, tolerance=1e-6):
     return np.array(unique)
 
 
-def valid_index(index, shape):
+def valid_index(index, shape) -> tuple:
```
Comment (ksunden) [inline WrightTools/kit/_array.py]: typing.Tuple?

File: WrightTools/kit/_calculate.py [hunk 9]
```diff
@@ -135,7 +138,7 @@ def nm_width(center, width, units="wn"):
     return red - blue
 
 
-def symmetric_sqrt(x, out=None):
+def symmetric_sqrt(x, out=None) -> np.ndarray:
```
Comment (ksunden) [inline WrightTools/kit/_calculate.py]: Union(np.ndarray, float, complex)

Either that or define a Generic x, and returns the type of x

File: WrightTools/kit/_ini.py [hunk 10]
```diff
@@ -55,12 +56,12 @@ def clear(self):
         self.config = configparser.ConfigParser()
 
     @property
-    def dictionary(self):
+    def dictionary(self) -> dict:
```
Comment (ksunden) [inline WrightTools/kit/_ini.py]: typing.Dict[str, dict]?

File: WrightTools/kit/_ini.py [hunk 11]
```diff
@@ -107,7 +108,7 @@ def has_section(self, section):
         self.config.read(self.filepath)
         return self.config.has_section(section)
 
-    def read(self, section, option):
+    def read(self, section, option) -> str:
```
Comment (ksunden) [inline WrightTools/kit/_ini.py]: Does not always return string, really quite a large union, honestly

**Comment by ksunden** [comment]
Test failures are due to the use of elipsis in the generic List, which does not accept such.

File: WrightTools/kit/_ini.py [hunk 12]
```diff
@@ -55,12 +56,12 @@ def clear(self):
...
 
     @property
-    def dictionary(self):
+    def dictionary(self) -> dict:
         """Get a python dictionary of contents."""
         self.config.read(self.filepath)
         return self.config._sections
 
-    def get_options(self, section):
+    def get_options(self, section) -> List[str, ...]:
```
[the remark this answers is not in the document]
Comment (ksunden) [reply WrightTools/kit/_ini.py]: In fact, it is an error

File: WrightTools/units.py [hunk 13]
```diff
@@ -158,7 +158,7 @@ def get_symbol(units):
         return kind(units)
 
 
-def get_valid_conversions(units):
+def get_valid_conversions(units) -> tuple:
```
[the remark this answers is not in the document]
Comment (p770193) [reply WrightTools/units.py]: I can do that--I wasn't sure if the tuple always contains strings--can it return `None`? 

If `valid` is always a tuple of strings, I should also type hint for `units.kind`

File: WrightTools/kit/_ini.py (hunk 11 above)
Comment (p770193) [reply WrightTools/kit/_ini.py]: ok, I'll just remove the hint

File: WrightTools/kit/_ini.py (hunk 10 above)
Comment (ksunden) [reply WrightTools/kit/_ini.py]: perhaps even specify the subdictionary type

File: WrightTools/units.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ksunden) [reply WrightTools/units.py]: As of #709, it returns an empty tuple if there is no valid conversion. Prior to that, it raised KeyError

**Comment by ksunden** [comment]
Something strange happening on appveyor, not sure what right now, seems to be infinite recursion in an import of skimage

File: WrightTools/data/_channel.py (hunk 1 above)
Comment (ksunden) [reply WrightTools/data/_channel.py]: Also:

https://www.python.org/dev/peps/pep-0484/#the-numeric-tower

**Comment by ksunden** [comment]
Tests are in the processes passing on appveyor, so if the fault persists, its in the latter half of the tests or in clean up afterwards

**Comment by ksunden** [comment]
Somehow appveyor is reporting success, though it is still producing almost 900k lines of output, most of which regard some infinite recursion...

**Comment by ksunden** [comment]
@p770193 may be something with the broader order of imports:

https://github.com/wright-group/WrightTools/blob/abfb980a09543c15baebcc9e31f353cd1fddee95/WrightTools/data/__init__.py#L13

**Comment by p770193** [comment]
```
Traceback (most recent call last):
  File "C:\Users\dandk\Anaconda3\lib\weakref.py", line 624, in _exitfunc
    f()
  File "C:\Users\dandk\Anaconda3\lib\weakref.py", line 548, in __call__
    return info.func(*info.args, **(info.kwargs or {}))
  File "C:\Users\dandk\OneDrive\Documents\GitHub\WrightTools\WrightTools\_group.py", line 246, in close
    from .collection import Collection
  File "C:\Users\dandk\OneDrive\Documents\GitHub\WrightTools\WrightTools\__init__.py", line 17, in <module>
    from . import artists
  File "C:\Users\dandk\OneDrive\Documents\GitHub\WrightTools\WrightTools\artists\__init__.py", line 5, in <module>
    from ._base import *
  File "C:\Users\dandk\OneDrive\Documents\GitHub\WrightTools\WrightTools\artists\_base.py", line 9, in <module>
    import matplotlib
  File "C:\Users\dandk\Anaconda3\lib\site-packages\matplotlib\__init__.py", line 131, in <module>
    from matplotlib.rcsetup import defaultParams, validate_backend, cycler
  File "C:\Users\dandk\Anaconda3\lib\site-packages\matplotlib\rcsetup.py", line 29, in <module>
    from matplotlib.fontconfig_pattern import parse_fontconfig_pattern
  File "C:\Users\dandk\Anaconda3\lib\site-packages\matplotlib\fontconfig_pattern.py", line 22, in <module>
    from pyparsing import (Literal, ZeroOrMore, Optional, Regex, StringEnd,
  File "C:\Users\dandk\Anaconda3\lib\site-packages\pyparsing.py", line 943, in <module>
    collections.MutableMapping.register(ParseResults)
  File "C:\Users\dandk\Anaconda3\lib\abc.py", line 158, in register
    if issubclass(subclass, cls):
```

Some tweaking of import order leads to the same loop being initiated by pyparsing via matplotlib, rather than the dask via skimage that we saw earlier

**Pull request merged**